### ISSUE TITLE
Using an alternative to COLMAP

### ISSUE THREAD

**Issue by henrypearce4D** [graphdeco-inria/gaussian-splatting#7]
Hi, datasets I am testing do not align well in colmap, I'm looking to see if there is a way to use an alternative such as Metashape or Reality Capture.

Have you considered accepting other formats such as metashape cameras.xml or reality capture cameras.csv?
Do you know of any conversation tools from these formats to colmaps model files (cameras.bin, images.bin, points3D.bin) ?

Or, could .nvm format be accepted?  metashape and colmap both export this.

**Comment by grgkopanas** [comment]
Unfortunately, we currently only support Colmap and json files similar to nerf-synthetic dataset, but if you have the time and energy we would love to help you create a pull request that supports new formats.

**Comment by henrypearce4D** [comment]
Hi George, can you send a link to an exmaple of the json file similar to nerf-synthetic dataset please?

**Comment by grgkopanas** [comment]
You can download them here:

https://www.matthewtancik.com/nerf

also the code for loading such files is here:
https://github.com/graphdeco-inria/gaussian-splatting/blob/main/scene/dataset_readers.py#L179
and for colmap files is here:
https://github.com/graphdeco-inria/gaussian-splatting/blob/main/scene/dataset_readers.py#L68

**Comment by Snosixtyboo** [comment]
> Hi George, can you send a link to an exmaple of the json file similar to nerf-synthetic dataset please?
> 
> would this be a dataset typically with;
> 
> ```
> transforms.json
> /images
> ```
> 
> Or something else?

Please note that one would have to adapt the json loading code to support general scenes. Right now it assumes the synthetic NeRF blender data set with 800x800 pixel resolution. I'll see if i can bump that for general transforms.json later today.

**Comment by Snosixtyboo** [comment]
> > Hi George, can you send a link to an exmaple of the json file similar to nerf-synthetic dataset please?
> > would this be a dataset typically with;
> > ```
> > transforms.json
> > /images
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Or something else?
> 
> Please note that one would have to adapt the json loading code to support general scenes. Right now it assumes the synthetic NeRF blender data set with 800x800 pixel resolution. I'll see if i can bump that for general transforms.json later today.

But the remaining problem with transforms.json would still be that there is agreed-upon way to reference input geometry (i.e., a point cloud), which we require for initialization. Adding support for Reality Capture would be better, but some effort. We have to check if we can find the time to do this.

**Comment by henrypearce4D** [comment]
Great thankyou both.

I can generate this .json from converting a metashape camera.xml, and can also export a scene camera matching .ply.
Could this file be referenced if placed in the dataset folder and referenced in some way if there is an agreed open naming convention?

The drums dataset is now training, it looks fantastic! the set doesn't have a point cloud file, is this not required for the synthetic blender datasets?

From looking at the transforms.json there is no camera intrinsics so I presume these data sets are rendered as 'pinhole' camera model type, for my own dataset I can test exporting undistorted images and remove the intrinsics from the transforms.json.

**Comment by Snosixtyboo** [comment]
> Great thankyou both.
> 
> I can generate this .json from converting a metashape camera.xml, and can also export a scene camera matching .ply. Could this file be referenced if placed in the dataset folder and referenced in some way if there is an agreed open naming convention?
> 
> The drums dataset is now training, it looks fantastic! the set doesn't have a point cloud file, is this not required for the synthetic blender datasets?
> 
> From looking at the transforms.json there is no camera intrinsics so I presume these data sets are rendered as 'pinhole' camera model type, for my own dataset I can test exporting undistorted images and remove the intrinsics from the transforms.json.

The blender dataset is simple enough that we can train it from random points, this will not work as well for arbitrary scenes (see the paper). Combining transforms and a local .ply file could work. One would have to remove the hard assumption of 800x800 pixels and look for a .ply in the local directory. Basically a combination of the methods that are already in the code base. I can check if I find the time to do this, but I'm not sure unfortunately.

**Comment by henrypearce4D** [comment]
Hi @Snosixtyboo, just bumping this question.

**Comment by Snosixtyboo** [comment]
Hi, 

we are looking at all issues in order of severity, and unfortunately, this isn't highest right now so it could take a while if we get round to it. But cameras other than pinhole will likely not be supported due to the fact that our code is rasterization based, which means that advanced lens distortions are not easily realized. This would be much easier in any ray-based  (NeRF) approach, but I don't know what MipNeRF360 or InstantNGPs support for arbitrary cameras is.

**Comment by henrypearce4D** [comment]
Ok thankyou, if general transforms will not be looked at soon I will continue investigating other methods.

**Comment by BennetLeff** [comment]
@grgkopanas Since you said "but if you have the time and energy we would love to help you create a pull request that supports new formats" I took a shot at it but it didn't work at all, haha! 

I'm loading a polycam dataset which is a `.ply` and similar data to the `transforms.json` file (plus some extra). I've had great success with Lidar with nerfs so I think it is worthwhile to investigate. A couple questions that I've run into:
How should I define my matrices? 
The following code is for the COLMAP importer. 
```
matrix = np.linalg.inv(np.array(frame["transform_matrix"]))
R = -np.transpose(matrix[:3,:3])
R[:,0] = -R[:,0]
T = -matrix[:3, 3]
``` 
Maybe it's right for a polycam import too? Polycam gives data like `t_00, t_10, t_01, ... t_22` that are entries to  `transform_matrix`. But I can't tell where they should go exactly (and if I need to process with transposes afterwards). I've tried using `i` as a row index and `j` as a column index but the nerfstudio's documentation on coordinate conventions is making me doubt that's correct. Any insight?

Another simpler question, is it important to calculate normals for each point? The polycam `ply` implementation doesn't have them. 

Thanks for everything! Any help fixing this hot-mess of Gaussian clouds would be great :)

**Comment by grgkopanas** [comment]
> Ok thankyou, if general transforms will not be looked at soon I will continue investigating other methods.

Other camera models can be supported during capture by letting COLMAP undistort the images by during calibration. Supporting other camera models in rendering can be challenging when you do rasterization.

**Comment by grgkopanas** [comment]
> @grgkopanas Since you said "but if you have the time and energy we would love to help you create a pull request that supports new formats" I took a shot at it but it didn't work at all, haha!
> 
> My first question before any follow up is: Is it important to calculate normals for each point? The polycam `ply` implementation that I'm trying to import doesn't have them.
> 
One of the  properties of our method is that exactly it doesnt need normals for any initial point clouds. I am not sure where not having normals appeared as a problem for you, can you please elaborate more?

> Thanks for everything! Any help fixing this hot-mess of Gaussian clouds would be great :)

**Comment by BennetLeff** [comment]
While I could initialize my point cloud randomly to start, I have a point cloud from my polycam capture that I imagine will aid the training. Based on my reading of the COLMAP datareader which I'm basing my datareader on I've run into the code [here](https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/scene/dataset_readers.py#L112) in `fetch_ply`. In my implementation I've had to set normals to a zero vector just to get the code to run but I can't imagine that's helpful. For what it's worth, I've gotten renderings that do 'match' my scene but with lots of artifacts. My guess is that some of the artifacts are related to the 'densification' process because my pointcloud goes from 34.5 MB to ~240 MB with many more points added but that's a complete guess. Additionally, rendering the initial PLY file and the PLY generated by the trained model shows an abundance of points placed outside the scene (outside the room I captured in).

**Comment by grgkopanas** [comment]
Normals in this case are completely irrelevant, you did well to just set
them to zero vectors. Often the quality of the results is correlated with
how a scene is captured, unfortunately the optimization can get stuck in
weird solutions similar to NeRF or other methods, if your results don't
look good it's probably because the algorithm could not find a good
solution given the cameras that were provided as input.

Most often the capture styles that work best are object-oriented where you
go around an object and capture images from all angles. Take a look as an
example how mipNeRF360 dataset was captured.

Best,
George

**Comment by BennetLeff** [comment]
I'll start with some smaller examples to test then, thanks!

**Comment by BennetLeff** [comment]
Here's an example I grabbed on my desk of my airpods case. In the SIBR viewer it seems like I've parsed some data wrong but I don't understand how. Any insight? 

[output.webm](https://github.com/graphdeco-inria/gaussian-splatting/assets/5445536/d9fe0c72-7671-42ba-9d87-c4824426395b)

**Comment by grgkopanas** [comment]
Depending on how comfortable you are with coding you might want to try to add this function in the "Scene" Object to try to visualize the point cloud and the cameras.

This will not work out of the box because its a very old implementation but you can get a rough idea or try to implement your own visualizer.

```
def debugView(self, scale=1.0):
        import polyscope as ps

        ps.init()

        ps_pc_diffuse = ps.register_point_cloud("point_cloud_diffuse", self.diffuse_point_cloud.global_xyz.detach().cpu().numpy(), enabled=True, radius=0.00007)
        ps_pc_diffuse.add_color_quantity("point_cloud_colors", self.diffuse_point_cloud.global_features[:,:3].detach().cpu().numpy(), enabled=True)

        cam_poses = np.array([x.camera_center.cpu().detach().numpy() for x in self.getTrainCameras(scale)]).squeeze()
        ps_cameras = ps.register_point_cloud("cameras", cam_poses, enabled=True)

        dirs = [geom_transform_vectors(torch.tensor([[0.0, 0.0, 1.0]]), x.world_view_transform.inverse().cpu()).cpu().detach().numpy() for x in self.getTrainCameras(scale)]
        ps_cameras.add_vector_quantity("cam_dirs", np.array(dirs).squeeze(), enabled=True)

        """
        poi = self.findCenterOfInterest(scale)
        print(poi)
        ps_poi = ps.register_point_cloud("point_of_interest", poi, enabled=True)
        """

        ps_x = ps.register_point_cloud("origin_x", np.array([[0., 0., 0.]]), enabled=True)
        ps_y = ps.register_point_cloud("origin_y", np.array([[0., 0., 0.]]), enabled=True)
        ps_z = ps.register_point_cloud("origin_z", np.array([[0., 0., 0.]]), enabled=True)
        ps_x.add_vector_quantity("dir_x", np.array([[1., 0., 0.]]), enabled=True, color=(1., 0., 0.))
        ps_y.add_vector_quantity("dir_y", np.array([[0., 1., 0.]]), enabled=True, color=(0., 1., 0.))
        ps_z.add_vector_quantity("dir_z", np.array([[0., 0., 1.]]), enabled=True, color=(0., 0., 1.))

        #ps.look_at(cam_poses[0], np.array(dirs).squeeze()[0])

        ps.show()
```

**Comment by BennetLeff** [comment]
Thanks I'll give something like that a shot. 

I've tried to play around with the transform matrix / `R`+`t` and I'm obviously doing something wrong because it stops rendering anything more than a blank screen. The COLMAP camera parser just reads in `R|t` from the ex/intrinsics - which gives me a blank screen. That makes sense because Polycam's coordinate convention is different - it's
> Y axis points up, and -Z axis points in the direction that the camera was pointing
So I tried flipping those axes but it still doesn't work. Oddly, I only just noticed I was doing something 'wrong'. What I accidentally originally copied was the blender importer which does
```
            matrix = np.linalg.inv(np.array(frame["transform_matrix"]))
            R = -np.transpose(matrix[:3,:3])
            R[:,0]=-R[:,0]
            T = matrix[:3, 3]
```

This however **does** converge / train. I don't understand why we might want to invert and transpose the matrix.

**Comment by grgkopanas** [comment]
You can ignore the transpose, its unnecessary, it actually gets nulled out by this transpose:
https://github.com/graphdeco-inria/gaussian-splatting/blob/main/utils/graphics_utils.py#L40

It should not be there and it's an embarrassing mishap that we have such a messy code regarding our camera models.

**Comment by Entongsu** [comment]
Hi, @BennetLeff. I am trying to use the camera transformation from Polycom to render the GS, but I have found that it returns a blank image. The camera transformation matrix may have some problems.  Would you know how to transform the camera matrix from Polycom to the GS rendering requirement? Thank you.
